File: agents/backend/onyx/server/features/addiction_recovery_ai/schemas/progress.py
"""
Pydantic schemas for progress tracking endpoints
"""

from typing import List, Optional, Dict
from pydantic import BaseModel, Field, validator
from datetime import datetime, date


class LogEntryRequest(BaseModel):
    """Request schema for logging daily entry"""
    user_id: str = Field(..., description="User ID")
    date: str = Field(..., description="Entry date (ISO format)")
    mood: str = Field(..., description="Mood level")
    cravings_level: int = Field(..., ge=0, le=10, description="Cravings level (0-10)")
    triggers_encountered: List[str] = Field(default_factory=list, description="Triggers encountered")
    consumed: bool = Field(default=False, description="Whether substance was consumed")
    notes: Optional[str] = Field(default=None, description="Additional notes")

    @validator('mood')
    def validate_mood(cls, v):
        """Validate mood value"""
        valid_moods = ['excellent', 'good', 'neutral', 'poor', 'terrible']
        if v.lower() not in valid_moods:
            raise ValueError(f"Invalid mood. Must be one of: {valid_moods}")
        return v.lower()


class LogEntryResponse(BaseModel):
    """Response schema for logged entry"""
    entry_id: str = Field(..., description="Entry ID")
    user_id: str = Field(..., description="User ID")
    date: str = Field(..., description="Entry date")
    mood: str = Field(..., description="Mood level")
    cravings_level: int = Field(..., description="Cravings level")
    triggers_encountered: List[str] = Field(default_factory=list, description="Triggers encountered")
    consumed: bool = Field(..., description="Whether substance was consumed")
    notes: Optional[str] = Field(default=None, description="Additional notes")
    logged_at: datetime = Field(default_factory=datetime.now, description="Log timestamp")


class ProgressResponse(BaseModel):
    """Response schema for user progress"""
    user_id: str = Field(..., description="User ID")
    days_sober: int = Field(..., ge=0, description="Days sober")
    total_entries: int = Field(..., ge=0, description="Total log entries")
    streak_days: int = Field(..., ge=0, description="Current streak in days")
    longest_streak: int = Field(..., ge=0, description="Longest streak in days")
    progress_percentage: float = Field(..., ge=0, le=100, description="Progress percentage")
    recent_entries: List[Dict] = Field(default_factory=list, description="Recent log entries")


class StatsResponse(BaseModel):
    """Response schema for user statistics"""
    user_id: str = Field(..., description="User ID")
    total_days: int = Field(..., ge=0, description="Total days tracked")
    days_sober: int = Field(..., ge=0, description="Days sober")
    relapse_count: int = Field(..., ge=0, description="Number of relapses")
    average_cravings: float = Field(..., ge=0, le=10, description="Average cravings level")
    most_common_triggers: List[str] = Field(default_factory=list, description="Most common triggers")
    trends: Dict = Field(default_factory=dict, description="Progress trends")


class TimelineResponse(BaseModel):
    """Response schema for progress timeline"""
    user_id: str = Field(..., description="User ID")
    timeline: List[Dict] = Field(default_factory=list, description="Timeline events")
    milestones: List[Dict] = Field(default_factory=list, description="Achieved milestones")
    relapses: List[Dict] = Field(default_factory=list, description="Relapse events")

File: addiction_recovery_ai/addiction_recovery_ai_mobile/src/types/progress.ts
import type { Mood } from './constants';

// Progress Types
export interface LogEntryRequest {
  user_id: string;
  date: string;
  mood: Mood;
  cravings_level: number; // 0-10
  triggers_encountered?: string[];
  consumed?: boolean;
  notes?: string;
}

export interface LogEntryResponse {
  entry_id: string;
  user_id: string;
  date: string;
  mood: string;
  cravings_level: number;
  triggers_encountered: string[];
  consumed: boolean;
  notes?: string;
  logged_at: string;
}

export interface ProgressResponse {
  user_id: string;
  days_sober: number;
  total_entries: number;
  streak_days: number;
  longest_streak: number;
  progress_percentage: number;
  recent_entries: LogEntryResponse[];
}

export interface StatsResponse {
  user_id: string;
  total_days: number;
  days_sober: number;
  relapse_count: number;
  average_cravings: number;
  most_common_triggers: string[];
  trends: Record<string, any>;
}

export interface TimelineResponse {
  user_id: string;
  timeline: TimelineEvent[];
  milestones: Milestone[];
  relapses: RelapseEvent[];
}

export interface TimelineEvent {
  date: string;
  type: string;
  description: string;
  data?: Record<string, any>;
}

export interface Milestone {
  id: string;
  date: string;
  title: string;
  description: string;
  days_sober: number;
}

export interface RelapseEvent {
  id: string;
  date: string;
  notes?: string;
  triggers?: string[];
}

